### DISCUSSION TITLE
Consider to support the pickle of MX, SX, DM

### DISCUSSION THREAD

**Discussion by eaglema** [casadi/casadi#3370] [Ideas]
Hi Joel and Joris,

I know CasADi allows the user to pickle CasADi function for parallel computation and so on, which has been very cool. I wonder whether it is possible to allow to pickle MX, SX and DM symbols in the future, which are not supported at present? For example, the following code will report an error:

x = casadi.SX.sym('x')
with open(r'C:\some_writable_folder\x_test.pkl', 'wb') as f: 
    pickle.dump(x, f)

With such functionality, the users will be able to pickle their own objects with both CasADi symbols and functions. This will make parallel computation easier.

Best wishes,
Yingjie

**Comment by jgillis** [comment]
There is something like this inplemented here: https://gitlab.kuleuven.be/meco-software/rockit/-/blob/master/rockit/ocp.py#L282

**Reply by eaglema** [reply]
Hi Joris,

I tried the functions and they worked very well! Thank you so much!

Best wishes,
Yingjie

**Reply by eaglema** [reply]
Hi Joris,

Sorry for bothering you! Do you have any idea about how to resolve the issue below? 

Best wishes,
Yingjie

**Comment by eaglema** [comment]
> There is something like this inplemented here: https://gitlab.kuleuven.be/meco-software/rockit/-/blob/master/rockit/ocp.py#L282

Hi Joris,

This works for the case when I only want to serialize and deserialize my object. But there is error reported when I try to combine the functions with the multiprocessing package in Python. For example, the following code doesn't work:

```
from multiprocessing import Pool
import casadi as cas
from utility import rockit_pickle_context, rockit_unpickle_context

class Foo:
    def __init__(self):
        self.x = cas.SX.sym('x')

    def sim_fun(self, p):
        y = self.x + p
        print('done')
        return y

if __name__ == '__main__':
    foo = Foo()
    with rockit_pickle_context():
        with Pool(5) as p:
            print(p.map(foo.sim_fun, [1, 2, 3]))
```

The error is like this:
in xxx file,, in sim_fun
    y = self.x + p
TypeError: unsupported operand type(s) for +: 'SX' and 'int'_

It seems during the parallelism, the normal operations (e.g., +, -, *, /) cannot be recognized anymore.

Best wishes,
Yingjie

**Reply by CheeGY** [reply]
Hi Yingjie,

Do you have any luck with pickling of MX, SX and DM symbols, and the multithreading package in Python?
